Issue #5605 unconsumed input on sendError

gregw opened the pull request:
Add Connection:close if content can't be consumed during a sendError. Processed after the request has returned to the container.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #5605 unconsumed input on sendError

Add Connection:close if content can't be consumed during a sendError. Processed after the request has returned to the container.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5605-unconsumed-send-error

gregw committed:
Update from review

 + Add close on all uncommitted requests when content cannot be consumed.

gregw commented:
@sbordet I'm now trying to consumeAll during the completion of any request without a committed response (ie when we can still add the Connection: close).

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -406,7 +410,13 @@ public boolean handle()
+
+                            // Add Connection:close if we can't consume the input
+                            if (!_request.getHttpInput().consumeAll())
+                                ensureConnectionClose();
```
Isn't this a duplication of what happens in `COMPLETE`?

gregw replied:
It is similar logic, but happens at a different time - ie between returning from a dispatch after a sendError and before an ERROR dispatch... which may do lots and lots of stuff before we ever get to COMPLETE.

sbordet replied:
@gregw and so, if we may dispatch to an error handler that then reads the request content, we should not do it ourselves here, no?

gregw replied:
I don't think that is a valid use case for an error handler to do.   I think it is far better we consume the input from the failed request and then do the ERROR dispatch.

sbordet replied:
All righty.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -492,10 +502,18 @@ public boolean handle()
+                                _response.sendError(HttpStatus.NOT_FOUND_404);
+                                break;
+                            }
+                            // If content has not been consumed and we can't consume it now without blocking, then
```
Remove "then" at end of line, it duplicates the one on the next line.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -545,6 +558,29 @@ public boolean handle()
+
+                return new HttpField(HttpHeader.CONNECTION, fields.stream()
+                    .flatMap(field -> Stream.of(field.getValues()))
+                    .collect(Collectors.joining(", ")));
```
Shouldn't duplicates be discarded?

gregw replied:
Perhaps... but I'm just really leaving the other values exactly as they are, other than amalgamating on to the same header.  If somebody else put a duplicate in there, then that's there problem and I don't want to waste cycles here fixing it.

sbordet replied:
Okay.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -545,6 +558,29 @@ public boolean handle()
+
+            return new HttpField(HttpHeader.CONNECTION, fields.stream()
+                .flatMap(field -> Stream.of(field.getValues()))
+                .collect(Collectors.joining(", ")) + ",close");
```
You are using ", " (comma+space) to delimit value, so add a space before "close" when you add it.
Same as above, should duplicates be discarded?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelState.java:
```diff
@@ -904,8 +904,6 @@ public void sendError(int code, String message)
                     throw new IllegalStateException(getStatusStringLocked());
             }
-            if (_outputState != OutputState.OPEN)
-                throw new IllegalStateException("Response is " + _outputState);
```
Why was this removed?

joakime replied:
I would also prefer a better error message ...

``` java
            if (_outputState != OutputState.OPEN)
                throw new IllegalStateException("Response is not open (was " + _outputState + "));
```

gregw replied:
See lines 895 and 896 above.... it was duplicated.

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/AsyncRequestReadTest.java:
```diff
@@ -328,6 +328,7 @@ public void testPartialReadThenClose() throws Exception
 
             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             assertThat(in.readLine(), containsString("HTTP/1.1 200 OK"));
+            assertThat(in.readLine(), containsString("Connection: close"));
```
Gah! This test will break as soon as we change the order of the headers, or if a JDK decides to change how things are stored in a `Map`. Not a biggie for this PR, but...

joakime replied:
Use our HttpTester here.

gregw replied:
yeah well it broke as soon as I changed our strategy for closing, so yes fragile.... but at least it made me look.

gregw replied:
Not keen on major rewrite of this test - there are lot of such tests.   We can fix them in another PR.

gregw committed:
Update from review

 + fixed comment
 + space comma

gregw committed:
Only consume input in COMPLETE if response is >=200 (ie not an upgrade or similar)

gregw commented:
@sbordet I had to update to only try to consume the input if the response is >= 200, otherwise it was trying to consume input for 101 upgrades etc.
Don't we have a HTTP/2 case where we do an upgrade, but not with a 101?  Is this going to break it?

gregw committed:
Updated to be less adventurous

I do not think it was valid to always consumeAll in COMPLETE as this could break upgrades with both 101s and 200s
Instead I have reverted to having this consumeAll logic only:
 + in sendError once control has passed back to the container and we are about to generate an error page.
 + in front of all the sendRedirection that we do without calling the application first.

Extra tests also added

gregw commented:
@sbordet @lorban I know you have approved, but can you re-review.

I've decided to go for a less general solution as I lost confidence that `consumeAll` was always appropriate, as it was breaking upgrade.
Instead I'm just doing in `sendError` and in the many places that we do a redirection ourselves without calling the application first.  This is a bit uglier, but I think it is safer.

lorban reviewed (approved):
You've got 1 broken test (`AsyncRequestReadTest.testPartialReadThenClose`) but it should be easy to fix, so this LGTM if fixing the test does not require non-test code changes (which I think is the case).

gregw committed:
Updated to be less adventurous

reverted test

joakime committed:
Testcase for odd sendError(400) issue.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Testcase for odd sendError(400) issue.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet committed:
Fix for odd sendError(400) issue.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Always try to consumeAll on all requests

gregw commented:
@sbordet can you check the latest commit that always calls consumeAll

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-5605-unconsumed-send-error

gregw committed:
Refinements after testing in 10

gregw committed:
Refinements after testing in 10

Fixed test

sbordet commented on jetty-jaspi/src/main/java/org/eclipse/jetty/security/jaspi/modules/FormAuthModule.java:
```diff
@@ -173,6 +173,7 @@ public AuthStatus validateRequest(MessageInfo messageInfo, Subject clientSubject
                     }
 
                     response.setContentLength(0);
+                    Request.getBaseRequest(request).getHttpChannel().ensureContentConsumedOrConnectionClose();
```
I would put this code _inside_ `sendRedirect()` to: A) avoid duplicating it all over the places and B) avoid `Request.getBaseRequest(request)` since I'm already inside the Jetty `Response` where I can easily access the `HttpChannel`.

gregw replied:
It can't go inside of `sendRedirect()` as it is perfectly OK for an application to send a redirect and then consume the content.
We can only do it on our own sendRedirect calls.  We could have our own method for `consumeAllAndSendRedirect`?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -492,10 +501,18 @@ public boolean handle()
+                            }
+
+                            // Indicate Connection:close if we can't consume all.
+                            if (_response.getStatus() >= 200)
```
I wonder whether this could be more lax... take into account 201 Created, 204 No Content, 206 Partial Content, etc.

gregw replied:
Why would 201, 204 and 206 not what this behaviour?   All we are doing is `consumeAll` a little earlier than we would have otherwise, so we get a chance to send a `Connection: close` if we can't consume all.     We just want to exclude partial responses (ie 100, 101 etc.) which handle content differently.

sbordet replied:
Doh, wrong comment on my side.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -511,12 +528,7 @@ public boolean handle()
-                        // TODO callback to do a non-blocking consumeAll at this point and only call completed when
-                        // TODO that is done.
-
+                        
```
The block above has a call to `abort()`, which then falls through, which I don't think it's right, we should `break`.

gregw replied:
Why is that not correct?   It still sets up the final call to completed.
Besides, that is existing behaviour and not really related to this PR.

sbordet replied:
Okay.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -406,7 +410,12 @@ public boolean handle()
+                            _response.setStatus(code);
+
+                            // Add Connection:close if we can't consume the input
+                            ensureContentConsumedOrConnectionClose();
```
I think this needs a comment on why it is done here, because otherwise it should be done only in `COMPLETE`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java:
```diff
@@ -408,28 +408,12 @@ public void onCompleted()
-                }
-            }
+            if (LOG.isDebugEnabled())
+                LOG.debug("{}unconsumed input {}", _parser.isChunking() ? "Possible " : "", this);
```
Perhaps simplify the log -- do we really need the "chunking" difference? Perhaps just log the `_parser` object.

sbordet reviewed (changes requested):
I think the handling of redirects must be reviewed, as we don't want to duplicate all the calls to `ensureContentConsumedOrConnectionClose()`.

gregw committed:
Fixed comment from review

gregw committed:
Updates from review

+ added redirect methods that consumeAll
+ ensureContentConsumedOrConnectionClose renamed to ensureConsumeAllOrNotPersistent
+ ensureConsumeAllOrNotPersistent now handles HTTP/1.0 and HTTP/1.1 differently

gregw committed:
better consumeAll implementation

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -497,6 +497,36 @@ public void sendProcessing() throws IOException
+    }
+
+    /**
+     * Sends a response with one of the 300 series redirection codes.
```
Fix javadocs: "Sends a response with the given redirection status code".

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -545,6 +562,60 @@ public boolean handle()
+
+                    return new HttpField(HttpHeader.CONNECTION, fields.stream()
+                        .flatMap(field -> Stream.of(field.getValues()))
+                        .collect(Collectors.joining(", ")) + ", " + HttpHeaderValue.CLOSE.asString());
```
Since many clients also add `Connection: keep-alive` is it worth to remove it even if the protocol is HTTP/1.1?

joakime replied:
There's also `Connection: upgrade`?

gregw replied:
removing keep-alive as well makes the code rather complex.  Close should always override keep-alive.... not so in our generator for HTTP/1.0, I will change that.

gregw replied:
OK I removed the `keep-alive` as well and added unit tests for all the variations... good thing too as I was adding an extra ',' in some circumstances.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java:
```diff
@@ -672,29 +672,44 @@ public boolean eof()
     public boolean consumeAll()
     {
-        synchronized (_inputQ)
+        while (true)
```
Add javadocs with the precise semantic about the return value.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -1240,10 +1240,10 @@ public boolean checkContext(final String target, final Request baseRequest, fina
+            baseRequest.getResponse().sendRedirect(
+                HttpServletResponse.SC_MOVED_TEMPORARILY,
+                baseRequest.getRequestURI() + (baseRequest.getQueryString() == null ? "/" : ("/?" + baseRequest.getQueryString())),
+                true);
```
Maybe use a local variable for the query?

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java:
```diff
@@ -233,6 +233,88 @@ public void test404HtmlAccept() throws Exception
+        assertThat(response.get(HttpHeader.CONTENT_TYPE), containsString("text/html;charset=ISO-8859-1"));
+        assertThat(response.getField(HttpHeader.CONNECTION), nullValue());
+        assertContent(response);
+    }
```
One more test with HTTP/1.1 with `Connection: keep-alive`?

sbordet replied:
Can we also add a test with gzip content? Can we assert that we did not ungzip it?

gregw replied:
@sbordet I've added `keep-alive` to the existing 1.1 test as it doesn't change behaviour, but shows that it is correctly ignored.
@joakime can I get you to do that gzip test?

gregw committed:
update from review

 + better javadoc
 + filter out keep-alive
 + added more tests

gregw committed:
update from review

 + better javadoc

gregw committed:
update from review

 + fixed form redirection test for http 1.0 and 1.1

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java:
```diff
@@ -659,16 +659,15 @@ else if (contentLength != field.getLongValue())
                             {
                                 close = true;
                                 _persistent = false;
                             }
```
Now you first check for keep-alive, but then if close is there, shouldn't you remove keep-alive?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -497,6 +497,36 @@ public void sendProcessing() throws IOException
+    {
+        if (consumeAll)
+            getHttpChannel().ensureConsumeAllOrNotPersistent();
         if ((code < HttpServletResponse.SC_MULTIPLE_CHOICES) || (code >= HttpServletResponse.SC_BAD_REQUEST))
```
I think there is a `HttpStatus.isRedirection(int)` that you can use here.

sbordet reviewed (changes requested):
A couple of improvements to do.

gregw committed:
update from review

 + HttpGenerator removes keep-alive if close present
 + Use isRedirection

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x-400-gzip' into jetty-9.4.x-5605-unconsumed-send-error
